datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaConnector.java

```diff
@@ -147,15 +150,32 @@ public synchronized void onAssignmentChange(List<DatastreamTask> tasks) {
       Set<DatastreamTask> toCancel = new HashSet<>(_runningTasks.keySet());
       toCancel.removeAll(tasks);
 
-      for (DatastreamTask task : toCancel) {
-        ConnectorTaskEntry connectorTaskEntry = _runningTasks.remove(task);
-        // Spawn a separate thread to attempt stopping the connectorTask. The connectorTask will be canceled if it
-        // does not stop within a certain amount of time. This will force cleanup of connectorTasks which take too long
-        // to stop, or are stuck indefinitely. A separate thread is spawned to handle this because the Coordinator
-        // requires that onAssignmentChange() must complete quickly, and will kill the assignment threads if they take
-        // too long.
+      // Spawn a separate thread to attempt stopping the connectorTask. The connectorTask will be canceled if it
+      // does not stop within a certain amount of time. This will force cleanup of connectorTasks which take too long
+      // to stop, or are stuck indefinitely. A separate thread is spawned to handle this because the Coordinator
+      // requires that onAssignmentChange() must complete quickly, and will kill the assignment threads if they take
+      // too long.
+      if (toCancel.size() > 0) {
+        List<Future<DatastreamTask>> stopTaskFutures = toCancel.stream()
+            .map(task -> _shutdownExecutorService.submit(() -> stopTask(task, _runningTasks.get(task))))
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+
         _shutdownExecutorService.submit(() -> {
-          stopTask(task, connectorTaskEntry);
+          List<DatastreamTask> toRemoveTasks = stopTaskFutures.stream().map(stopTaskFuture -> {
+            try {
+              return stopTaskFuture.get();
+            } catch (ExecutionException | InterruptedException e) {
+              e.printStackTrace();
+            }
+            return null;
+          }).filter(Objects::nonNull).collect(Collectors.toList());
+
+          if (toRemoveTasks.size() > 0) {
+            synchronized (_runningTasks) {
+              toRemoveTasks.forEach(_runningTasks::remove);
+            }
+          }
```

Comment:
Earlier we had an issue with tasks that weren't stopped properly but were removed from running tasks. Now such tasks won't be removed. Other than logging that an attempt to stop a task failed or timed out, do we have an idea on how to deal with such "zombie" tasks? If they'll be living in-memory indefinitely, can they be a source for a memory leak?

Reply:
In the "restartDeadTasks" method we check again is a given task is "dead" and if yes it will attempt to stop it again. So if its a transient error eventually it should be stopped and removed. At least this is my understanding, pls correct me if I am missing something.

Reply:
For non-transient errors, we will never be able to start the task to replace this dead task which isn't able to stop. This can lead to stopped partitions problems and will be very hard to debug and RC.

Please ensure that you add a metric to track when we are unable to stop a task so what we can quickly check for this metric when we run into issues. Can you perhaps also add a comment explaining that this can lead to stopped partitions issue?

Also, please ensure that we fix pieces of code that don't honor the interrupt signal in the future so that we can minimize this scenario. @surajkn and I don't have the full context on what kind of exceptions are seen such that we are unable to stop tasks. We are also not aware of how often his happens, so any insights from @vmaheshw or @jzakaryan will be helpful here.